enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,342 @@
+2. Create a `Service` to expose the webhook pods, with generated TLS cert/key pairs.
+3. Create a `ConfigMap` where the service serving CA can be injected.
+4. Create a `MutatingWebhookConfiguration` that fires the webhook for all created pods.
+5. When the service serving CA in item 3 is injected or updated, the operator needs to update the
+   caBundle provided to the webook configuration in 4.
+
+[1]
+https://kubernetes.io/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/#mutating-webhook-configuration
+
+#### Bundle API
+
+```yaml
+kind: ClusterBundle
+apiVersion: subscription.openshift.io/v1alpha1
+metadata:
+  name: <name>
+  ... # standard metadata, cluster-scoped so no namespace
+spec:
+  entitlement:
+    name: <secret name with entitlement keys, must be in openshift-config namespace>
+  yumRepository:
+    name: <configMap name with yum repo definition, must be in openshift-config namespace>
+  aggregateToClusterRoles: # cluster roles which the ClusterBundle RBAC should aggregate to.
+                           # This can override the operator default
+  - edit
+  - admin
+status:
+  conditions:
+  - type: Invalid
+    status: "False" # standard k8s conditions, should obey the "abnormal-true" convention
+  - ... # additional informational conditions
+  clusterRole: <name>-subscription-bundle # clusterRole is cluster-scoped
+  shares:
+  - name: <name>-subscription-entitlement # shares are cluster-scoped
+  - name: <name>-subscription-yumRepository
+```
+
+### Risks and Mitigations
+
+**Risk:** Entitlement keys can leak outside of the `openshift-config` namespace to unauthorized
+workloads.
+
+_Mitigation:_ Access to the entitlement keys is gated by the cluster role generated for the
+projected resource shares associated with the bundle. Cluster admins are responsible for either
+
+1. Creating a `ClusterRoleBinding` for the service accounts that need access to the entitlement, OR
+2. Declaring cluster roles to aggregate the `ClusterBundle` to via the `aggregateToClusterRoles`
+   array, OR
+3. Configuring the operator to aggregate the generated `ClusterRole` to appropriate system roles by
+   default.
+
+**Risk:** Shares and cluster roles are orphaned if the subscription `ClusterBundle` is deleted.
+
+_Mitigation:_ Owner references are used to ensure proper cleanup. Note that `ClusterBundle` is
+cluster-scoped, so it can an create owner references on any object.
+
+## Design Details
+
+### Open Questions
+
+1. Should a subscription bundle only allow one reference for `entitlement` and `yumRepository`?
+   Can/should we allow multiple?
+2. Should we restrict entitlement info to live in `openshift-config`? Should we allow entitlements
+   and yum repo info to live in any namespace?
+3. Will this work for multi-tenant clusters? Should we consider APIs that are namespace-scoped?
+4. Should we inject a subscription bundle into every pod (and all containers within the pod) by
```

Comment:
i think these questions need at least a placeholder answer before we can merge this and expect it to be implementable.

1) allowing multiple will likely save you pain down the road.  it's not hard to imagine multiple yumRepositories being needed
2) restricting to openshift-config seems reasonable until you get to multitenant.  probably better to support arbitrary namespaces up front.
3) see above, doesn't seem like it works for multitenant clusters (at least not in a way where the tenants can control their own content.  It would have to all be managed by a single admin who could access everyone's stuff)
4) my inclination is no we should not inject by default, because if you do it by default, how does someone opt-out of it?  But it might be reasonable for the subscription bundle to have a field that says "injectAlways" or something, so that an admin could make it so all pods get it, if the admin so chooses.

Reply:
> allowing multiple will likely save you pain down the road. it's not hard to imagine multiple yumRepositories being needed

Agree - will update the API.

> restricting to openshift-config seems reasonable until you get to multitenant. probably better to support arbitrary namespaces up front.

Agree as well - will update the `ClusterBundle` to allow `Secrets`/`ConfigMap` refs in any namespace.

> see above, doesn't seem like it works for multitenant clusters (at least not in a way where the tenants can control their own content. It would have to all be managed by a single admin who could access everyone's stuff)

Ack on this, too. Will add the `Bundle` API which in all likelihood can be released first.

> But it might be reasonable for the subscription bundle to have a field that says "injectAlways" or something, so that an admin could make it so all pods get it, if the admin so chooses

IMO this is a configuration option that is applied to the webhook, and should be exposed via an operator API. I don't think we _need_ this right away, but we could certainly consider an enhancement.

My understanding of OLM operators is that we will need a CRD to say "install this webhook", so we can add this option as a future enhancement.

Reply:
making it a config on the webhook would imply it's set for all bundles or none of them?  vs on a per bundle basis?

or maybe i'm not familiar enough w/ the webhook config/operator api

Reply:
My thought on injecting a cluster bundle everywhere:

1. We can inject everywhere via the "webhook" CRD to say "inject this cluster bundle into all pods." Configuration is gated by whomever can edit the webhook's CRD. We can use the API to say that you are only allowed to inject one bundle everywhere.
2. We can use a label or API field on `ClusterBundle` to let admins declare "inject this cluster bundle everywhere". For this to work we would need to allow multiple bundles to be injected. If we are able to support multiple entitlement keys and yum repo files, supporting multiple bundles isn't that much of a stretch imo.

My understanding is that you'd prefer something like option 2?

Reply:
correct, though i understand injecting multiple bundles is likely problematic since it results in potentially multiple mounts to the same path of a given pod.

but i really don't like the idea that this would be a config on the webhook.  I'd prefer the webhook not require separate configuration (or any configuration)

Reply:
so i think the question is, given we also said elsewhere that we want to allow multiple bundles, how do we actually inject multiple bundles into a pod?  is there actually a reasonable way to do it?  if not we might be back to saying one bundle only per pod.

at which point it becomes a question of how to enforce that when we also have the ability for a given bundle to say "always inject me"

Reply:
Thoughts on how to inject multiple bundles:

1. Allow a comma-separated list of bundles: `subscription.openshift.io/inject-bundle: rhel,jboss`
2. Entitlements cannot have overlapping keys in their `Secrets` - if keys overlap deny admission
3. The optional yum repos likewise cannot have overlapping keys in the `ConfigMaps` - if keys overlap deny admission

We would then use key-to-path mappings to ensure the keys and yum repo files are mounted correctly. For the namespace-scoped `Bundle` object, we can use a projected volume [1]. I believe rules 2 and 3 would also need to apply to a single `Bundle` which references multiple entitlements - we would need a sane way of merging these sets.

[1] https://kubernetes.io/docs/concepts/storage/volumes/#projected

Reply:
sounds reasonable to me.  i'd be comfortable only supporting a single bundle in the first pass since the api can easily be expanded to allow comma separated lists in the future.